lib/ansible/modules/system/osx_defaults.py

```diff
@@ -15,413 +16,975 @@
+        float,
+        (text_type, binary_type),
+        datetime.datetime,
+        Data,
+        list,
+        dict,
+    )
+    for value_type in supported_types:
+        if isinstance(value1, value_type) and isinstance(value2, value_type):
+            return True
+    return False
 
-        # Ensure the value is the correct type
-        self.value = self._convert_type(self.type, self.value)
 
-    # /init --------------------------------------------------------------- }}}
+class Data(binary_type):
+    """
+    Object representing binary data.
+    Instance should be initialize with binary data encoded in base64 codec.
+    """
+    # List of all Base64 accepted characters.
+    BASE64_CHARS = string.ascii_letters + string.digits + '+/='
 
-    # tools --------------------------------------------------------------- {{{
-    """ Converts value to given type """
+    # List of all text characters.
+    TEXT_CHARS = bytearray(set([7, 8, 9, 10, 12, 13, 27]) | set(range(0x20, 0x100)) - set([0x7f]))
 
-    def _convert_type(self, type, value):
+    def __new__(cls, data):
+        if isinstance(data, cls):
+            return data
 
-        if type == "string":
-            return str(value)
-        elif type in ["bool", "boolean"]:
-            if isinstance(value, (binary_type, text_type)):
-                value = value.lower()
-            if value in [True, 1, "true", "1", "yes"]:
-                return True
-            elif value in [False, 0, "false", "0", "no"]:
-                return False
-            raise OSXDefaultsException("Invalid boolean value: {0}".format(repr(value)))
-        elif type == "date":
+        # Try to convert unicode to ascii string.
+        if isinstance(data, text_type):
             try:
-                return datetime.datetime.strptime(value.split("+")[0].strip(), "%Y-%m-%d %H:%M:%S")
-            except ValueError:
-                raise OSXDefaultsException(
-                    "Invalid date value: {0}. Required format yyy-mm-dd hh:mm:ss.".format(repr(value))
-                )
-        elif type in ["int", "integer"]:
-            if not str(value).isdigit():
-                raise OSXDefaultsException("Invalid integer value: {0}".format(repr(value)))
-            return int(value)
-        elif type == "float":
-            try:
-                value = float(value)
-            except ValueError:
-                raise OSXDefaultsException("Invalid float value: {0}".format(repr(value)))
-            return value
-        elif type == "array":
-            if not isinstance(value, list):
-                raise OSXDefaultsException("Invalid value. Expected value to be an array")
-            return value
+                data = data.encode('ascii')
+            except UnicodeEncodeError:
+                pass
 
-        raise OSXDefaultsException('Type is not supported: {0}'.format(type))
+        # Check if data is a valid base64 string. Short strings are not
+        # considered as binary.
```

Comment:
Because with a small dataset, it's hard to know with a good certitude if it's binary data or not (check next comment fro the "fox" example). The more data you have, the more chance you have to hit non-text characters.

The 52 value is a bit arbitrary but I found that you rarely have useful binary data that are less then 37 bytes long (52 bytes b64 data converted to binary will gives 37 bytes). For example a 8 by 8 pixels 8 bit images will take 64 bytes. I did some tests on real data in the context of the macOS preferences to find this "52". I wish there was a better solution but I didn't found one.